### PR TITLE
feat: NFTAccess: implementation of AllowlistHook

### PR DESCRIPTION
**Pull request by davidprepo**
https://www.notion.so/prepo/NFTAccessHook-SC-Mini-Spec-5bbf9f36b33f4005933272429b274740#f5f925b0fb7d42e4a67282952fb84809

This is the head PR for the NFT access list feature.

Implemented as one PR, as per previous comments of skeleton implementations being too skinny.

### CODE REVIEW

**Review by ramenforbreakfast — APPROVED** [review]
So I made quite a few changes here since the submission as it was would not have worked.

- Defined an `internal` `_setAllowlist` that an inheriting contract would call.
- Moved out `SafeAccessControlEnumerable` from `AllowlistHook` since we should allow the inheriting contract to define access control.
- Added the role to `DepositHook` and setup the overriding `external` function, verifying that it reverts if not role holder.

I acknowledge that this could have been moved to `prepo-shared-contracts`, but this will have to happen in a later PR, because it uses `IAccountList` which was regrettably not already moved to `prepo-shared-contracts`, so it would require modifying a bunch of contract imports.

File: apps/smart-contracts/core/test/AllowlistHook.test.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  let user: SignerWithAddress
+  let allowlist: SignerWithAddress
+
+  beforeEach(async () => {
+    ;[deployer, user, allowlist] = await ethers.getSigners()
+    allowlistHook = await allowlistHookFixture()
+  })
+
+  describe('# initialize', () => {
+    it('should be initialized with no allowlist set', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowlistHook.test.ts:21]:
```suggestion
    it('initializes allowlist to zero address', async () => {
```

File: apps/smart-contracts/core/test/AllowlistHook.test.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      expect(await allowlistHook.getAllowlist()).to.eq(ethers.constants.AddressZero)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowlistHook.test.ts:22]: are we phasing out the `ZERO_ADDRESS` constant? if using this form, can we make it more concise (i.e. without the `ethers.constants.` each time)?

File: apps/smart-contracts/core/test/AllowlistHook.test.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    it('is idempotent for non-zero address', async () => {
+      await allowlistHook.setAllowlist(allowlist.address)
+      expect(await allowlistHook.getAllowlist()).to.eq(allowlist.address)
+
+      await allowlistHook.setAllowlist(allowlist.address)
+
+      expect(await allowlistHook.getAllowlist()).to.eq(allowlist.address)
+    })
+
+    it('is idempotent for zero address', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowlistHook.test.ts:53]:
```suggestion
    it('is idempotent if zero address', async () => {
```

File: apps/smart-contracts/core/test/AllowlistHook.test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    it('sets to zero address', async () => {
+      await allowlistHook.setAllowlist(allowlist.address)
+      expect(await allowlistHook.getAllowlist()).to.eq(allowlist.address)
+
+      await allowlistHook.setAllowlist(ethers.constants.AddressZero)
+
+      expect(await allowlistHook.getAllowlist()).to.eq(ethers.constants.AddressZero)
+    })
+
+    it('is idempotent for non-zero address', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/AllowlistHook.test.ts:44]:
```suggestion
    it('is idempotent if non-zero address', async () => {
```

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

File: apps/smart-contracts/core/test/AllowlistHook.test.ts (hunk 2 above)
Comment (ramenforbreakfast) [reply apps/smart-contracts/core/test/AllowlistHook.test.ts:22]: No, we are not, I missed this, I think i commented on some other PRs that we should stick to ZERO_ADDRESS

**Pull request merged**